File: packages/types/schema.py
"""
Shared result schema for Raven Brain Analytics.
Used by both the FastAPI backend and inference worker.
"""
from __future__ import annotations

from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# --------------------------------------------------------------------------
# Enums
# --------------------------------------------------------------------------

class JobStatus(str, Enum):
    queued = "queued"
    processing = "processing"
    completed = "completed"
    failed = "failed"


class CognitiveLoad(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class EngagementZone(str, Enum):
    high = "high"
    medium = "medium"
    low = "low"
    drop = "drop"


class EmotionLabel(str, Enum):
    excitement = "excitement"
    curiosity = "curiosity"
    surprise = "surprise"
    neutral = "neutral"
    boredom = "boredom"
    confusion = "confusion"


# --------------------------------------------------------------------------
# Sub-models
# --------------------------------------------------------------------------

class InputMetadata(BaseModel):
    filename: str
    size_bytes: int
    duration_seconds: Optional[float] = None
    fps: Optional[float] = None
    resolution: Optional[str] = None
    upload_url: Optional[str] = None


class BrainRegionActivation(BaseModel):
    nacc: float = Field(..., ge=0.0, le=1.0, description="Nucleus Accumbens - reward/anticipation")
    mpfc: float = Field(..., ge=0.0, le=1.0, description="Medial Prefrontal Cortex - valuation")
    ains: float = Field(..., ge=0.0, le=1.0, description="Anterior Insula - emotional salience")
    visual_cortex: float = Field(..., ge=0.0, le=1.0, description="Visual cortex intensity")
    amygdala: float = Field(..., ge=0.0, le=1.0, description="Amygdala - emotional response")


class EmotionalArc(BaseModel):
    second: float
    valence: float = Field(..., ge=-1.0, le=1.0)
    arousal: float = Field(..., ge=0.0, le=1.0)
    dominant_emotion: EmotionLabel


class RetentionPoint(BaseModel):
    second: float
    retention_pct: float = Field(..., ge=0.0, le=100.0)


class EngagementZoneMarker(BaseModel):
    start_seconds: float
    end_seconds: float
    zone: EngagementZone
    label: str
    score: float = Field(..., ge=0.0, le=1.0)


class PacingMetrics(BaseModel):
    avg_scene_duration: float
    scene_count: int
    visual_change_rate: float
    pacing_score: float = Field(..., ge=0.0, le=1.0)
    rhythm_consistency: float = Field(..., ge=0.0, le=1.0)


class BenchmarkComparison(BaseModel):
    category: str
    percentile: int = Field(..., ge=0, le=100)
    avg_attention_in_category: float
    your_attention: float
    top_10_pct_threshold: float
    sample_size: int


class HookAnalysis(BaseModel):
    scroll_stop_probability: float = Field(..., ge=0.0, le=1.0)
    hook_strength: float = Field(..., ge=0.0, le=1.0)
    time_to_peak_ms: int
    first_frame_score: float = Field(..., ge=0.0, le=1.0)
    recommendation: str


class SummaryMetrics(BaseModel):
    predicted_attention: float = Field(..., ge=0.0, le=1.0)
    peak_activation: float = Field(..., ge=0.0, le=1.0)
    peak_segment: int = Field(..., ge=0)
    cta_strength: float = Field(..., ge=0.0, le=1.0)
    cognitive_load: CognitiveLoad
    predicted_segments: int = Field(..., ge=1)
    avg_valence: float = Field(0.0, ge=-1.0, le=1.0)
    avg_arousal: float = Field(0.0, ge=0.0, le=1.0)
    predicted_retention_pct: float = Field(0.0, ge=0.0, le=100.0)
    virality_score: float = Field(0.0, ge=0.0, le=1.0)
    overall_grade: str = Field("B")


class SectionScore(BaseModel):
    name: str
    start_seconds: float
    end_seconds: Optional[float] = None
    score: float
    rank: int


class TimelinePoint(BaseModel):
    segment_index: int
    second: float
    activation: float
    brain_regions: Optional[BrainRegionActivation] = None


class Insight(BaseModel):
    type: str
    text: str
    confidence: Optional[float] = None
    icon: Optional[str] = None


class Artifact(BaseModel):
    name: str
    url: str
    mime_type: str


class ModelMetadata(BaseModel):
    model_name: str
    model_version: str
    inference_time_ms: Optional[float] = None
    device: Optional[str] = None


class AnalysisResult(BaseModel):
    job_id: str
    status: JobStatus
    input: InputMetadata
    summary: SummaryMetrics
    sections: List[SectionScore]
    timeline: List[TimelinePoint]
    insights: List[Insight]
    artifacts: List[Artifact]
    model_meta: ModelMetadata
    created_at: str
    completed_at: Optional[str] = None
    brain_regions_summary: Optional[BrainRegionActivation] = None
    emotional_arc: Optional[List[EmotionalArc]] = None
    retention_curve: Optional[List[RetentionPoint]] = None
    engagement_zones: Optional[List[EngagementZoneMarker]] = None
    pacing: Optional[PacingMetrics] = None
    benchmark: Optional[BenchmarkComparison] = None
    hook_analysis: Optional[HookAnalysis] = None


# --------------------------------------------------------------------------
# Request / Response for API
# --------------------------------------------------------------------------

class CreateJobRequest(BaseModel):
    object_key: str
    filename: str
    size_bytes: int
    sections_config: Optional[Dict[str, Any]] = None


class JobResponse(BaseModel):
    job_id: str
    status: JobStatus
    created_at: str
    updated_at: str
    result_url: Optional[str] = None
    error: Optional[str] = None


class UploadSignResponse(BaseModel):
    upload_url: str
    object_key: str
    expires_in: int

File: apps/web/src/types/schema.ts
// Shared TypeScript types -- mirrors packages/types/schema.py

export type JobStatus = 'queued' | 'processing' | 'completed' | 'failed'
export type CognitiveLoad = 'low' | 'medium' | 'high'
export type EngagementZone = 'high' | 'medium' | 'low' | 'drop'
export type EmotionLabel = 'excitement' | 'curiosity' | 'surprise' | 'neutral' | 'boredom' | 'confusion'

export interface InputMetadata {
  filename: string
  size_bytes: number
  duration_seconds?: number
  fps?: number
  resolution?: string
  upload_url?: string
}

export interface BrainRegionActivation {
  nacc: number
  mpfc: number
  ains: number
  visual_cortex: number
  amygdala: number
}

export interface EmotionalArc {
  second: number
  valence: number
  arousal: number
  dominant_emotion: EmotionLabel
}

export interface RetentionPoint {
  second: number
  retention_pct: number
}

export interface EngagementZoneMarker {
  start_seconds: number
  end_seconds: number
  zone: EngagementZone
  label: string
  score: number
}

export interface PacingMetrics {
  avg_scene_duration: number
  scene_count: number
  visual_change_rate: number
  pacing_score: number
  rhythm_consistency: number
}

export interface BenchmarkComparison {
  category: string
  percentile: number
  avg_attention_in_category: number
  your_attention: number
  top_10_pct_threshold: number
  sample_size: number
}

export interface HookAnalysis {
  scroll_stop_probability: number
  hook_strength: number
  time_to_peak_ms: number
  first_frame_score: number
  recommendation: string
}

export interface SummaryMetrics {
  predicted_attention: number
  peak_activation: number
  peak_segment: number
  cta_strength: number
  cognitive_load: CognitiveLoad
  predicted_segments: number
  avg_valence: number
  avg_arousal: number
  predicted_retention_pct: number
  virality_score: number
  overall_grade: string
}

export interface SectionScore {
  name: string
  start_seconds: number
  end_seconds?: number
  score: number
  rank: number
}

export interface TimelinePoint {
  segment_index: number
  second: number
  activation: number
  brain_regions?: BrainRegionActivation
}

export interface Insight {
  type: string
  text: string
  confidence?: number
  icon?: string
}

export interface Artifact {
  name: string
  url: string
  mime_type: string
}

export interface ModelMetadata {
  model_name: string
  model_version: string
  inference_time_ms?: number
  device?: string
}

export interface AnalysisResult {
  job_id: string
  status: JobStatus
  input: InputMetadata
  summary: SummaryMetrics
  sections: SectionScore[]
  timeline: TimelinePoint[]
  insights: Insight[]
  artifacts: Artifact[]
  model_meta: ModelMetadata
  created_at: string
  completed_at?: string
  brain_regions_summary?: BrainRegionActivation
  emotional_arc?: EmotionalArc[]
  retention_curve?: RetentionPoint[]
  engagement_zones?: EngagementZoneMarker[]
  pacing?: PacingMetrics
  benchmark?: BenchmarkComparison
  hook_analysis?: HookAnalysis
}

export interface CreateJobRequest {
  object_key: string
  filename: string
  size_bytes: number
  sections_config?: Record<string, any>
}

export interface JobResponse {
  job_id: string
  status: JobStatus
  created_at: string
  updated_at: string
  result_url?: string
  error?: string
}

export interface UploadSignResponse {
  upload_url: string
  object_key: string
  expires_in: number
}
